hazelcast/src/hazelcast/client/metrics.cpp

```diff
@@ -0,0 +1,317 @@
+}
+
+probe_unit metric_descriptor::unit() const {
+    return unit_;
+}
+
+int metrics_dictionary::get_dictionary_id(const std::string &word) {
+    if (word.size() > MAX_WORD_LENGTH) {
+        throw std::invalid_argument("too long value in metric descriptor");
+    }
+
+    auto word_position = word_to_id.find(word);
+
+    if (word_position == word_to_id.end()) {
+        const int next_id = static_cast<int>(word_to_id.size());
+        word_to_id[word] = next_id;
+
+        return next_id;
+    }
+    else {
+        return word_position->second;
+    }
+}
+
+metrics_dictionary::const_iterator metrics_dictionary::begin() const noexcept {
+    return word_to_id.cbegin();
+}
+
+metrics_dictionary::const_iterator metrics_dictionary::end() const noexcept {
+    return word_to_id.cend();
+}
+
+std::size_t metrics_dictionary::size() const noexcept {
+    return word_to_id.size();
+}
+
+void output_buffer::write(byte val) {
+    buffer_.push_back(val);
+}
+
+void output_buffer::write(int32_t val) {
+    auto pos = buffer_.size();
+    buffer_.resize(pos + sizeof(int32_t));
+
+    boost::endian::endian_store<int32_t, sizeof(int32_t), boost::endian::order::big>(
+      buffer_.data() + pos, val);
+}
+
+void output_buffer::write(int64_t val) {
+    auto pos = buffer_.size();
+    buffer_.resize(pos + sizeof(int64_t));
+
+    boost::endian::endian_store<int64_t, sizeof(int64_t), boost::endian::order::big>(
+      buffer_.data() + pos, val);
+}
+
+void output_buffer::write(double val) {
+    auto pos = buffer_.size();
+    buffer_.resize(pos + sizeof(double));
+
+    boost::endian::endian_store<double, sizeof(double), boost::endian::order::big>(
+      buffer_.data() + pos, val);
+}
+
+void output_buffer::write(const std::string &str) {
+    for (char c: str) {
+        buffer_.push_back(static_cast<byte>(0));
+        buffer_.push_back(c);
+    }
+}
+
+void output_buffer::write(const std::vector<byte> &vec) {
+    buffer_.insert(buffer_.end(), vec.begin(), vec.end());
+}
+
+const std::vector<byte> &output_buffer::content() const {
+    return buffer_;
+}
+
+void metrics_compressor::add_long(const metric_descriptor &descriptor, int64_t value) {
+    write_descriptor(descriptor);
+    metrics_buffer_.write(static_cast<byte>(VALUE_TYPE_LONG));
+    metrics_buffer_.write(value);
+}
+
+void metrics_compressor::add_double(const metric_descriptor &descriptor, double value) {
+    write_descriptor(descriptor);
+    metrics_buffer_.write(static_cast<byte>(VALUE_TYPE_DOUBLE));
+    metrics_buffer_.write(value);
+}
+
+std::vector<byte> metrics_compressor::get_blob() {
+    write_dictionary();
+    
+    std::vector<byte> compressed_dictionary = zlib_compress(dictionary_buffer_.content());
+    std::vector<byte> compressed_metrics = zlib_compress(metrics_buffer_.content());
+
+    output_buffer blob;
+
+    blob.write(static_cast<byte>((BINARY_FORMAT_VERSION >> BITS_IN_BYTE) & BYTE_MASK));
+    blob.write(static_cast<byte>(BINARY_FORMAT_VERSION & BYTE_MASK));
```

Comment:
if BINARY_FORMAT_VERSION is defined as byte, you would not need to`&` with `BYTE_MASK `, right?

Reply:
BINARY_FORMAT_VERSION is an int.

Reply:
does it have to be int? I look at used places and byte seems to be enough.

Reply:
It's `int` in Java and 2 bytes are reserved to serialize it in the binary format. So theoretically it can be up to 2^16-1. Anyways, changed it to byte.